Add ICorProfiler API to allow profiler created providers to have a callback when a session enables them

davmason opened the pull request:
With the current APIs there is no way for a profiler created EventPipe provider to have a callback when a session enables them. This plumbs through the existing EventPipe callback infrastructure.

CC @wiktork @jander-msft

davmason committed:
Provider enabled callbacks

davmason committed:
fix linux test build

davmason committed:
hopefully actually fix it

davmason committed:
No, like really fix it

Update eventpipewritingprofiler.cpp

noahfalk commented on src/tests/profiler/native/eventpipeprofiler/eventpipewritingprofiler.cpp:
```diff
@@ -188,6 +211,34 @@ HRESULT STDMETHODCALLTYPE EventPipeWritingProfiler::JITCachedFunctionSearchFinis
+    }
+
+    // The diagnostics client library sets keywords to 0xF00000000000 by default
+    if (match_any_keywords != 0xF00000000000
```
Does this mean keywords are also set when delivering the disable notification? That seems surprising although not necessarily problematic.

Is there particular behavior we are promising when a provider has been enabled more than once, or when we are disabling one session for a provider but other sessions are still running? Historically these behaviors have been poorly specified and it might be a lot more awkward to change the behavior in the future vs. ensuring it is what we want now.

davmason replied:
The current behavior is `is_enabled` will be true if any session has the provider active, and keywords will be whatever the combined keywords for all sessions are (on enable and disable).

This behavior matches ETW as far as I can tell

noahfalk reviewed (approved):
This basically looks good to me, though I am hoping we can better nail down the expected behavior of the callback when multiple sessions are being used concurrently.

davmason committed:
Update guids.cpp

davmason committed:
Checkin

davmason committed:
Update test and add comment explaining behavior

davmason commented:
Now that the change to make EventPipe callbacks match ETW is in I just rebased against main and updated the test plus added comments explaining the behavior in corprof.idl.

davmason committed:
Update guids.cpp
